### PR TITLE
MM-57319 Direct Message modal message improvement

### PR DESCRIPTION
**Pull request by bruno-keiko**
#### Summary
Issue #27175. In the Direct Message modal, once the user selects 7 members, the help text below the field is updated to “You can't add more than 7 people. Please create a channel to include more people.” The text ‘create a channel’ is a link and clicking on it dismisses this modal and triggers the Create channel modal.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/27175
https://mattermost.atlassian.net/browse/MM-57319

#### Screenshots
Before:
Screenshot from 2024-05-31 12-43-55
After:
DMmodal

#### Release Note
UI change. Direct message modals help text changed
```release-note
Updated help text in Direct Messages modal
```

### CODE REVIEW

File: server/config.mk [hunk 1]
```diff
@@ -13,7 +13,7 @@
 ENABLED_DOCKER_SERVICES ?= mysql postgres inbucket
 
 # Disable entirely the use of docker
-MM_NO_DOCKER ?= false
+MM_NO_DOCKER ?= true
```
Comment (larkox) [inline server/config.mk:16]: Revert this unrelated change

File: webapp/channels/src/components/more_direct_channels/list/index.ts [hunk 2]
```diff
@@ -127,6 +131,14 @@ export function makeGetOptions(): (state: GlobalState, users: UserProfile[], val
     );
 }
 
+function mapDispatchToProps(dispatch: Dispatch) {
+    return {
+        actions: bindActionCreators({
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/index.ts:136]: Since the component is a function component, I would prefer to use the hook (`useDispatch`) directly. But 0/5

@hmhealey Thoughts here?

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 3]
```diff
@@ -105,13 +124,21 @@ const List = React.forwardRef((props: Props, ref?: React.Ref<MultiSelect<OptionV
...
+                        values={{
+                            num: MAX_SELECTABLE_VALUES - props.values.length,
+                        }}
+                    />
+                    {'Please '}
+                    <span
+                        onClick={handleCreateChannel}
+                        className='link'
+                    >{'create a channel'}</span>
+                    {' to include more people.'}
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:140]: Use formatted Message for this, so it is translatable. You can use the chunks functionality to add the link (there are more examples in the code).

File: webapp/channels/src/i18n/en.json [hunk 4]
```diff
@@ -4203,7 +4203,7 @@
   "multiselect.maxGroupMembers": "No more than 256 members can be added to a group at once.",
   "multiselect.numGroupsRemaining": "Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {group} other {groups}}. ",
   "multiselect.numMembers": "{memberOptions, number} of {totalCount, number} members",
-  "multiselect.numPeopleRemaining": "Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. ",
+  "multiselect.numPeopleRemaining": "You can't add more than 7 people. Please {createChannel} to include more people.",
```
Comment (larkox) [inline webapp/channels/src/i18n/en.json:4206]: Looking at the code, you are not changing the default value of this string, but adding a new one. So this change should not be needed (the new string would have to be added here, though (to do that run `npm run i18n-extract`)).

**Review by larkox — CHANGES_REQUESTED** [review]
Great job. A few changes.

**Comment by begench-g** [comment]
> Great job. A few changes.

Hello @larkox. Can you review it again? I fixed. Thank you for the review!

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 5]
```diff
@@ -107,11 +124,19 @@ const List = React.forwardRef((props: Props, ref?: React.Ref<MultiSelect<OptionV
             numRemainingText={
                 <FormattedMessage
                     id='multiselect.numPeopleRemaining'
-                    defaultMessage='Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. '
+                    defaultMessage={'Use ↑↓ to browse, ↵ to select. You can\'t add more than 7 people. Please <a>create a channel</a> to include more people. '}
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:127]: The way this is implemented right now, we show this always. The idea is to show this only when we have reached the maximum number of people, right?

**Review by larkox — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 5 above)
Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx:127]: Yes, you're right. I apologize for the misunderstanding. I will fix it.

Comment (larkox) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx:127]: You can make up a new id. I guess `multiselect.maxPeople` should work.

Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: Thank you @larkox! I did it. Is it good practice or should I change something?

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 6]
```diff
@@ -122,21 +145,7 @@ const List = React.forwardRef((props: Props, ref?: React.Ref<MultiSelect<OptionV
...
-                            return (
-                                <a onClick={handleCreateChannel}>
-                                    {chunks}
-                                </a>
-                            );
-                        },
-                    }}
-                />
-
+                remaingText
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:148]: Is this working? The way this is right now, you are passing the function, and I would expect you to call the function and show the result instead.

Maybe it works because it considers it a function component? 0/5. Can you verify?

Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: I did it.

Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: @larkox  can you review it

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 7]
```diff
@@ -36,108 +41,160 @@ type Props = {
...
-            handlePageChange={props.handlePageChange}
-            handleInput={props.search}
-            handleDelete={props.handleDelete}
-            handleAdd={props.addValue}
-            handleSubmit={props.handleSubmit}
-            noteText={note}
-            maxValues={MAX_SELECTABLE_VALUES}
-            numRemainingText={
+
+        const RemainingText = () => {
```
Comment (hmhealey) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:115]: Defining this here sort of as a component within a component looks a bit odd to me because it's accessing the props and context of the outer component. It feels like something that may break at some point.

Would you mind changing this so that it's not a functional component? Something like
```js
let remainingText;
if (props.values.length >= MAX_SELECTABLE_VALUES) {
   remainingtext = ...;
} else {
   remainingtext = ...;
}
```
should work the same, but it wouldn't share props between the components

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 8]
```diff
@@ -36,108 +41,160 @@ type Props = {
...
-                    defaultMessage='Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. '
+                    id='multiselect.maxPeople'
+                    defaultMessage={
+                        "Use ↑↓ to browse, ↵ to select. You can't add more than 7 people. Please <a>create a channel</a> to include more people. "
+                    }
                     values={{
-                        num: MAX_SELECTABLE_VALUES - props.values.length,
+                        a: (chunks: React.ReactNode) => {
+                            return (
+                                <a onClick={handleCreateChannel}>{chunks}</a>
```
Comment (hmhealey) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:138]: This link needs an href for accessibility. Without that, it won't be selectable by users using the tab key.

With that, `handleCreateChannel` will also need to call `preventDefault()` on the mouse event that's passed to it

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR!

As mentioned above, would you mind reverting the stylistic changes here? I assume your editor did them automatically since they look like the changes Prettier makes.

Also, could you fix the broken unit tests like I mentioned on our community server?

**Comment by begench-g** [comment]
> @begench-g There are several style changes that seem unneeded specially in list.tsx, which make the review quite complicated to track what actually changed. Do you mind reverting the pure stylistic changes?

@larkox . Thank you for your help. I have completed the task. Can you please review it again?

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 8 above)
Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: @hmhealey  Thank you for your review. I fixed it. Could you review it?

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 7 above)
Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: @hmhealey  Thank you for your review. I fixed it. Could you review it?

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 9]
```diff
@@ -82,6 +94,13 @@ const List = React.forwardRef((props: Props, ref?: React.Ref<MultiSelect<OptionV
         }
     }
 
+    let remainingText;
+    if (MAX_SELECTABLE_VALUES >= props.values.length) {
+        remainingText = 'Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. ';
+    } else {
+        remainingText = 'Use ↑↓ to browse, ↵ to select. You can\'t add more than 7 people. Please <a>create a channel</a> to include more people. ';
+    }
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:102]: Changing just the default message doesn't work here. Let's use the FormattedMessage here directly. Also, it will avoid any weirdness in case the library complains about values being passed and not used.
```suggestion
    if (MAX_SELECTABLE_VALUES >= props.values.length) {
        remainingText = (
            <FormattedMessagedefineMessage
                id={'multiselect.numPeopleRemaining'}
                defaultMessage={'Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. '}
                value={{
                    num: MAX_SELECTABLE_VALUES - props.values.length,
                }}
            />
        );
    } else {
        remainingText = (
            <FormattedMessagedefineMessage
                id={'multiselect.maxPeople'}
                defaultMessage={'Use ↑↓ to browse, ↵ to select. You can\'t add more than 7 people. Please <a>create a channel</a> to include more people.'}
                value={{
                    a: (chunks: React.ReactNode) => {
                        return (
                            <a
                                href='#'
                                onClick={(e) => handleCreateChannel(e)}
                            >{chunks}</a>
                        );
                    },
                }}
            />
        );
    }
```

**Review by larkox — CHANGES_REQUESTED** [review]
One minor change more.

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 9 above)
Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: @larkox  I fixed it

File: webapp/channels/src/components/more_direct_channels/list/list.tsx [hunk 10]
```diff
@@ -96,9 +96,32 @@ const List = React.forwardRef((props: Props, ref?: React.Ref<MultiSelect<OptionV
...
+                    num: MAX_SELECTABLE_VALUES - props.values.length,
+                }}
+            />
+        );
     } else {
-        remainingText = 'Use ↑↓ to browse, ↵ to select. You can\'t add more than 7 people. Please <a>create a channel</a> to include more people. ';
+        remainingText = (
+            <FormattedMessage
+                id={'multiselect.maxPeople'}
+                defaultMessage={'Use ↑↓ to browse, ↵ to select. You can\'t add more than 7 people. Please <a>create a channel</a> to include more people.'}
```
Comment (larkox) [inline webapp/channels/src/components/more_direct_channels/list/list.tsx:112]: Sorry for all the back and forth. Can you make one small change more? I don't expect we ever change the value of `MAX_SELECTABLE_VALUES`, but just in case, change the 7 here for a variable (`{num}` instead of `7`) and pass it directly the `MAX_SELECTABLE_VALUES` in the values. That way, if we ever change it, we don't need to update the translations.

**Review by larkox — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 10 above)
Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: I completed the task, but I found another bug. When I try to close the direct message modal using dispatch, I can't find a way to do it; so I used props instead. However, this approach failed in testing.

**Review by larkox — APPROVED** [review]
Thank you very much!

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 10 above)
Comment (larkox) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: @hmhealey Any idea why dispatch may fail here?

Comment (begench-g) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: Because the "show" state of the MoreDirectChannels modal is in the component, not in the global store.

**Review by hmhealey — APPROVED** [review]
Thanks for making those changes. This looks good to me!

File: webapp/channels/src/components/more_direct_channels/list/list.tsx (hunk 10 above)
Comment (hmhealey) [reply webapp/channels/src/components/more_direct_channels/list/list.tsx]: I'd assume that it's because the modal's identifier is different than the one it was opened with. This way of closing the modal is actually preferable though because it's a bit simpler. The only reason you should really need to dispatch `closeModal` is if something outside of the modal can close it

**Comment by begench-g** [comment]
> @begench-g can you please check the CI failures?

I made updates to the snapshot test and it passed the CI test.

**Review by yasserfaraazkhan — APPROVED** [review]
